compiler/next/include/chpl/AST/BaseAST.h

```diff
@@ -0,0 +1,217 @@
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_AST_BASEAST_H
+#define CHPL_AST_BASEAST_H
+
+#include "chpl/AST/ASTTag.h"
+#include "chpl/AST/ID.h"
+#include "chpl/AST/ASTTypes.h"
+#include "chpl/Util/memory.h"
+
+#include <cassert>
+
+namespace chpl {
+namespace ast {
+
+/**
+  This is the base class for AST types.
+
+  Every AST class has:
+   * a tag (indicating which AST class it is)
+   * an ID (a sort of scoped location used as a key in maps)
+   * a list of child Exprs
+
+  The list of child Exprs is included in BaseAST to allow for
+  generic tree traversals of the AST.
+
+  someAst->isCallExpr() / someAst->toCallExpr() are available
+  and generated for all AST types.
+
+  std::less is defined for every AST class and it compares IDs.
+
+ */
+class BaseAST {
+ friend class Builder;
+
+ private:
+  asttags::ASTTag tag_;
+  ID id_;
+  ASTList children_;
+
+  /**
+    This function needs to be defined by subclasses.
+    It should check only those fields defined in subclasses
+    (it should not check the BaseAST fields tag, ID, or children).
+    It can assume that other has the same type as the receiver.
+   */
+  virtual bool contentsMatchInner(const BaseAST* other) const = 0;
+
+ protected:
+  BaseAST(asttags::ASTTag tag);
+  BaseAST(asttags::ASTTag tag, ASTList children);
+  // called by the Builder
+  void setID(ID id) { id_ = id; }
+
+ public:
+  virtual ~BaseAST() = 0; // this is an abstract base class
+
+  /**
+    Returns the tag indicating which BaseAST subclass this is.
+   */
+  asttags::ASTTag tag() const {
+    return tag_;
+  }
+
+  /**
+    Returns the ID of this AST node.
+   */
+  ID id() const {
+    return id_; 
+  }
+
+  /**
+    Returns the number of child AST nodes in the tree directly under this one.
+   */
+  int numChildren() const {
+    return children_.size();
+  }
+  /**
+    Returns the i'th child AST node in the tree directly under this one.
+    This function returns a "borrow" of the AST node. It is managed
+    by this object.
+   */
+  const BaseAST* child(int i) const {
+    assert(0 <= i && i < children_.size());
+    return children_[i].get();
+  }
+
+  /**
+    Returns 'true' if this symbol contains another AST node.
+    This is an operation on the IDs.
+   */
+  bool contains(const BaseAST* other) const {
+    return this->id_.contains(other->id_);
+  }
+
+  bool shallowMatch(const BaseAST* other) const;
+  /*
+  bool contentsMatch(const BaseAST* other) const;
```

Comment:
I need to just remove the comment. It's something I used to have but no longer need. I'm trying to minimize the number of virtual methods required (since each new AST class has to define those). For matching, the individual AST classes need only override `contentsMatchInner`.